core/src/avm2/optimize.rs

```diff
@@ -419,1179 +574,1466 @@ pub fn optimize<'gc>(
         }
     }
 
-    let has_simple_scoping = has_simple_scope_structure(code, &jump_targets, method_body);
+    let has_simple_scoping = has_simple_scope_structure(code, &jump_targets, method_exceptions);
 
-    // TODO: Fill out all ops, then add scope stack and stack merging, too
-    let mut state_map: HashMap<i32, Locals<'gc>> = HashMap::new();
+    // Map of op index -> local types + stack heights
+    let mut state_map: HashMap<i32, (Locals<'gc>, usize, usize)> = HashMap::new();
+    let mut expected_stack_heights: HashMap<i32, usize> = HashMap::new();
+    let mut expected_scope_stack_heights: HashMap<i32, usize> = HashMap::new();
```

Comment:
No, `expected_stack_heights` is also inserted into the target of a jump where the stack height is known before the jump. We don't actually know the stack contents at the target of the jump, and `state_map` should eventually change so that it holds a Stack and ScopeStack, not just the stack heights as usizes.